Enable "Run Query in New Tab" in SQL Lab

nickbarnwell opened the pull request:
Fixes #1147 – saw this was on the roadmap but it's important for the workflow I personally use so worst case I'll abandon this patch if there's already a fix in mainline.

Commit 7d33c8f shows the first approach I took, before reading a bit more about Redux; I can rebase it away and add tests if the approach in f3326a3 looks good.

n.b. This was my first time working with React/Redux, apologies in advance if this isn't idiomatic; I tried my best to base it on the rest of the codebase.

Commit:
Enable "Clone to New Tab" btn in QueryHistoryTable

Method #1; doesn't feel very clean.

Going to attempt to reimplement using an action and changing state directly
through the reducer rather than creating a new QueryEditor object directly from
the QueryTable

Commit:
Move Clone Logic to Action

mistercrunch commented on caravel/assets/javascripts/SqlLab/reducers.js:
```diff
@@ -34,6 +34,23 @@ export const sqlLabReducer = function (state, action) {
+        dbId: (action.query.dbId) ? action.query.dbId : null,
+        schema: (action.query.schema) ? action.query.schema : null,
+        autorun: true,
+        sql: `${action.query.sql}`,
```
same as  `sql: action.query.sql,`

mistercrunch reviewed (changes requested):
One minor nit, but otherwise LGTM. I'm glad to see that people out there are already using SQL Lab. Thank you for the PR.

Commit:
Implement PR feedback

nickbarnwell commented:
Thanks for the quick feedback! Definitely should've caught that nit, cargo-culted a little too hard there.

Will add tests tonight; since I was already in there I added a few for other reducer actions around QueryEditors - would you prefer those be separated out into another PR?

Commit:
Clean up reducer action; fix bug

Bug => Attempting to clone anything other than the most recent Query for a given
TabbedQueryEditor would throw an exception, because we depended on lastQueryId
to find the title of the QueryEditor to clone. Since you can only activate a
clone from the currently active tab, we can instead fetch the ID of the Editor
to copy the Title of from the tip of tabHistory.

Commit:
Tests for Reducer Action

nickbarnwell commented on caravel/assets/javascripts/SqlLab/reducers.js:
```diff
@@ -34,6 +34,20 @@ export const sqlLabReducer = function (state, action) {
+        sql: action.query.sql,
+      };
+
+      return sqlLabReducer(state, actions.addQueryEditor(qe));
```
Wasn't super DRY essentially copying addQueryEditor into this action. This seems better, but not terribly clean either…suggestions for improvement welcome!

nickbarnwell commented on caravel/assets/spec/javascripts/sqllab/reducers_spec.js:
```diff
@@ -0,0 +1,30 @@
+import { describe, it } from 'mocha';
+import { expect } from 'chai';
+
+describe('sqlLabReducer', () => {
```
Couldn't find anything to model these tests off of, let me know how you'd prefer them to be structured.

nickbarnwell commented on caravel/assets/spec/javascripts/sqllab/reducers_spec.js:
```diff
@@ -0,0 +1,30 @@
+    });
+
+    it('should prefix the new tab title with "Copy of"', () => {
+      expect(newState.queryEditors[1].title).to.include("Copy of");
```
Assuming that our new queryEditor will always be the last element is all well and good, but seems terribly hacky to me. Any suggestions or are you fine with it?

nickbarnwell commented on caravel/assets/spec/javascripts/sqllab/reducers_spec.js:
```diff
@@ -0,0 +1,30 @@
+
+    it('should have at most one more tab', () => {
+      expect(newState.queryEditors).have.length(2);
+      //expect(newState.queryEditors.filter((qe) => qe.sql == testQuery.SQL).length).to.equal(2)
```
Woops. Will remove this when I clean up any other nits from code review.

nickbarnwell reviewed (commented):
There are a few places where I wasn't sure how you prefer things to be tested and/or implemented. Called them out to save you some work when reviewing.

Commit:
Fix CodeClimate feedback

mistercrunch commented:
Thanks for the PR, this looks good to me as is, I'll do a blitz soon to get some test coverage on SQL Lab, but for now this is good!
